clippy_lints/src/missing_fields_in_debug.rs

```diff
@@ -0,0 +1,276 @@
+    /// }
+    /// impl fmt::Debug for Foo {
+    ///     fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+    ///         formatter
+    ///             .debug_struct("Foo")
+    ///             .field("data", &self.data)
+    ///             .finish_non_exhaustive()
+    ///     }
+    /// }
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub MISSING_FIELDS_IN_DEBUG,
+    pedantic,
+    "missing fields in manual `Debug` implementation"
+}
+declare_lint_pass!(MissingFieldInDebug => [MISSING_FIELDS_IN_DEBUG]);
+
+fn report_lints(cx: &LateContext<'_>, span: Span, span_notes: Vec<(Span, &'static str)>) {
+    span_lint_and_then(
+        cx,
+        MISSING_FIELDS_IN_DEBUG,
+        span,
+        "manual `Debug` impl does not include all fields",
+        |diag| {
+            for (span, note) in span_notes {
+                diag.span_note(span, note);
+            }
+            diag.help("consider including all fields in this `Debug` impl")
+                .help("consider calling `.finish_non_exhaustive()` if you intend to ignore fields");
+        },
+    );
+}
+
+/// Checks if we should lint in a block of code
+///
+/// The way we check for this condition is by checking if there is
+/// a call to `Formatter::debug_struct` or `Formatter::debug_tuple`
+/// and no call to `.finish_non_exhaustive()`.
+fn should_lint<'tcx>(
+    cx: &LateContext<'tcx>,
+    typeck_results: &TypeckResults<'tcx>,
+    block: impl Visitable<'tcx>,
+) -> bool {
+    let mut has_finish_non_exhaustive = false;
+    let mut has_debug_struct_tuple = false;
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        let ExprKind::MethodCall(path, recv, ..) = &expr.kind else {
+            return ControlFlow::Continue(());
+        };
+        let recv_ty = typeck_results.expr_ty(recv).peel_refs();
+
+        if [sym::debug_struct, sym::debug_tuple].contains(&path.ident.name)
+            && match_type(cx, recv_ty, &paths::FORMATTER)
+        {
+            has_debug_struct_tuple = true;
+        }
+
+        if path.ident.name.as_str() == "finish_non_exhaustive" && match_type(cx, recv_ty, &paths::DEBUG_STRUCT) {
+            has_finish_non_exhaustive = true;
+        }
+        ControlFlow::Continue(())
+    });
+
+    !has_finish_non_exhaustive && has_debug_struct_tuple
+}
+
+/// Attempts to find unused fields assuming that the item is a struct
+fn check_struct<'tcx>(
+    cx: &LateContext<'tcx>,
+    typeck_results: &TypeckResults<'tcx>,
+    block: &'tcx Block<'tcx>,
+    self_ty: Ty<'tcx>,
+    item: &'tcx Item<'tcx>,
+    data: &VariantData<'_>,
+) {
+    let mut field_accesses = FxHashSet::default();
+
+    let _: Option<!> = for_each_expr(block, |expr| {
+        if let ExprKind::Field(target, ident) = expr.kind
+            && let target_ty = typeck_results.expr_ty(target).peel_refs()
+            && target_ty == self_ty
+        {
+            field_accesses.insert(ident);
+        }
+        ControlFlow::Continue(())
+    });
```

Comment:
It would be good to also catch things like `.field("some_field", &())` or where the value is computed without accessing the field